rfcs/detecting_sensitive_textual_content/20230308-implementation_plan_filtering_and_designating_results_with_sensitive_textual_content.md

```diff
@@ -0,0 +1,458 @@
+                                    "query": f'"{term}"',
+                                    "fields": ["tags.name", "title", "description"],
+                                }
+                            }
+                            for term in SENSITIVE_TERMS
+                        ]
+                    }
+                },
+            },
+            "dest": {"index": destination_index},
+        },
+        slices="auto",
+        wait_for_completion=True,
+    )
+
+    self.refresh(index_name=destination_index, change_settings=True)
+
+    self.ping_callback()
+```
+
+Please note the following important details:
+
+1. `wait_for_completion` defaults to `True`, but it's worth explicitly including
+   as we rely on the fact that the request will block the process until the
+   reindexing is finished so that we know when we can promote the index.
+1. `slices` is set to `"auto"` so that Elasticsearch is free to decide the
+   optimal number of slices to use to parallelise the reindex.
+1. The source index name is not passed explicitly to the API: instead the method
+   must build it. This means we match the API of other calls (making updates to
+   the data refresh DAG much simpler) and can avoid needing to parse the model
+   name from the source index name for the call to `index_settings`.
+1. We explicitly refresh the index after it is finished reindexing to prime it
+   to be searched. ES only automatically refreshes indexes that are being
+   actively used, and the index should be refreshed before it starts being used.
+
+The creation of the new index follow the same pattern of creating the regular
+index: create the index, copy the data into the index, then issue the command to
+point the alias to the new index. This process is encapsulated in two actions:
+the `create_and_populate_filtered_index` action described above, and the
+existing `point_alias` action.
+
+> **Note**
+>
+> While both processes have a "copy" step, they are notably different. The
+> origin index receives its documents from Postgres via ingestion workers, which
+> push the documents into Elasticsearch. The filtered index receives its
+> documents via the reindex API and does not require the ingestion workers.
+
+### Airflow data refresh DAG factory (overview step 2)
+
+Adding these new steps to the Airflow data refresh DAG is simple and only
+requires adding the two additional actions to the
+[existing `action_data_map`](https://github.com/WordPress/openverse-catalog/blob/4b5811aa240ba3ee2d92adea3a8abf88eb1e8aa4/openverse_catalog/dags/data_refresh/data_refresh_task_factory.py#L182)
+with the target alias configured on the `point_alias` action, just as it is with
+the regular `promote` action.
+
+```py
+action_data_map: dict[str, dict] = {
+    "ingest_upstream": {},
+    "promote": {"alias": target_alias},
+    "create_and_populate_filtered_index": {},
+    "point_alias": {"alias": f"{target_alias}-filtered"}
+}
+```
+
```

Comment:
Checking I understand:  any changes to the `SENSITIVE_TERMS` list (addition or removal of terms) will not be reflected in search results until the next data refresh, right?

I'm hoping the data refresh will become much more stable soon so this won't be a problem🤞, but if we needed to it sounds like we could also manually trigger just the relevant `create_and_populate_filtered_index` and `point_alias` steps on the ingestion server to refresh these results without a full data refresh.

Do you think it would be worth having the steps to create the filtered index in a separate DAG? We could then trigger the DAG from the data refresh using a [TriggerDagRunOperator](https://docs.astronomer.io/learn/cross-dag-dependencies#triggerdagrunoperator) so that it would continue to be an automatic part of standard data refreshes, but we could also run it independently. We _would_ need to make sure the new DAG can't be manually triggered while a data refresh is already running, which would require a bit of tweaking.

Reply:
Exactly, I remembered talking about a one-week timeframe being acceptable at some point, but the instability of the data refresh might change that 😞 

> To prevent them running at the same time, is it a matter of adding some sort of lock in Airflow variables? I vaguely recall you'd done some work on another DAG to prevent this as well.

Yes, the data refresh DAGs themselves use a `data_refresh` pool with one slot and a custom `SingleRunExternalDAGSensor` to make sure that only one data refresh DAG runs at a time. We could add these new DAGs to the list of "data refresh DAGs", but that wouldn't be quite right by default -- because the 'parent' DAG will still be running when the filtered index creation DAG starts.

I think the `TriggerDagRunOperator` will allow us to [pass in conf options](https://registry.astronomer.io/providers/apache-airflow/versions/latest/modules/triggerdagrunoperator) when we trigger the filtered index creation DAG, so we should be able to pass a "force" parameter like you suggested. Then we would probably want a new, simple sensor that would `wait` as long as the data refresh is running, but also accept the "force" parameter as an override.

There might be a simpler way to do it, which I can look into, but I think that ought to work!

Reply:
> Then we would probably want a new, simple sensor that would wait as long as the data refresh is running

The data refresh takes so long: would it make more sense for the dag to fail if called during the data refresh rather than waiting? If the only time that kind of error can happen is if we manually call it (this dag itself won't be scheduled, right?) then waiting would be more confusing to me rather than an error that says "I can't do this while the data refresh is going".

Reply:
For sure, we could definitely fail the DAG instead. Now you mention it, the data refresh will include the filtered index creation step as is... so there shouldn't theoretically be any reason to wait and re-do it after the data refresh finishes, anyway. That sounds great 👍